automate.py

```diff
@@ -15,51 +16,49 @@ class Automate:
 
     def __init__(self, tracer):
         self._tracer = tracer
-        self._circuits = []
-        self.find_circuits()
         self.ids = set()
         self.scheduler = Scheduler()
 
     # pylint: disable=too-many-nested-blocks, too-many-branches
     def find_circuits(self):
         """Discover all circuits in a topology.
 
-        Using the list of flows per switch, run control plane
+        Using the list of stored flows from flow_manager, run control plane
         traces to find a list of circuits."""
         all_flows = {}
         circuits = []
+        stored_flows = get_stored_flows()
 
         for switch in self._tracer.controller.switches.copy().values():
             all_flows[switch] = []
             if switch.ofp_version == '0x04':
                 controller_port = Port13.OFPP_CONTROLLER
 
             try:
-                for flow in switch.generic_flows:
+                for flow in stored_flows[switch.dpid]:
+                    flow = flow['flow']
+                    if 'match' not in flow:
+                        continue
                     action_ok = False
                     in_port_ok = False
-                    if 'in_port' in flow.match and flow.match['in_port'] != 0:
+                    if 'in_port' in flow['match'] \
+                            and flow['match']['in_port'] != 0:
                         in_port_ok = True
                     if in_port_ok:
-                        for action in flow.actions:
-                            if action.action_type == 'output' \
-                                    and action.port != controller_port:
+                        for action in flow['actions']:
+                            if action['action_type'] == 'output' \
+                                    and action['port'] != controller_port:
                                 action_ok = True
                     if action_ok:
                         all_flows[switch].append(flow)
             except AttributeError:
                 pass
-
         for switch, flows in all_flows.items():
             for flow in flows:
-                in_port = flow.match['in_port']
+                in_port = flow['match']['in_port']
                 vlan = None
-                if 'vlan_vid' in flow.match:
-                    vlan = flow.match['vlan_vid']
-                if switch.ofp_version == '0x04':
-                    in_port = in_port.value
-                    if vlan:
-                        vlan = vlan.value
+                if 'vlan_vid' in flow['match']:
```

Comment:
@gretelliz, I think this `'vlan_vid'` will never enter this branch, could you double check since we're not doing fields mapping anymore? I think it should be `'dl_vlan'`

Reply:
On an `action_type` yes, but here it's on a `match`. Let me close this since you've fixed it.